### ISSUE TITLE
on_disconnect websocket handler is not called for all possible disconnects

### ISSUE THREAD

**Issue by martonilles** [mirumee/ariadne#578]
Currently `on_disonnect` callback is only called while handling `GQL_CONNECTION_TERMINATE` message, but the websocket can be disconnected in other places:
- exception in `handle_websocket_connection_init_message`
- on WebSocketDisconnect in `websocket_server`

Calling on_disconnect in those cases would be very useful to properly handle websocket related cleanups

**Comment by kissgyorgy** [comment]
This is a little bit painful to us too, here is our workaround:

```
class GraphQLApp(GraphQL):
    async def websocket_server(self, websocket: WebSocket) -> None:
        await super().websocket_server(websocket)
        await db.aclose()
```

with a huge comment block why it's necessary here, and why we can't use `on_disconnect`.

**Comment by rafalp** [comment]
@kissgyorgy it feels wrong to open/close service conenctions in websocket hooks. Every ASGI lib I've seen so far created connections pool on startup and picked connection from this pool on demand. Connections were closed only on app's shutdown.

**Comment by kissgyorgy** [comment]
We have a multi-tenant architecture, we can't have a pool, because the database is determined on connection from the client, so every client has a different connection which have to be closed when the websocket connection is closed, otherwise we would have hundreds of connections lingering unnecessarily. (Terrible legacy architecture, but we inherited it, not easy to change.)

**Comment by rafalp** [comment]
Almost a year later I am wondering if solution to this isn't as simple as moving `on_disconnect` logic to `finally` clause in `websocket_server`:

```python

    async def websocket_server(self, websocket: WebSocket) -> None:
        subscriptions: Dict[str, Subscription] = {}
        await websocket.accept("graphql-ws")
        try:
            while WebSocketState.DISCONNECTED not in (
                websocket.client_state,
                websocket.application_state,
            ):
                message = await websocket.receive_json()
                await self.handle_websocket_message(message, websocket, subscriptions)
        except WebSocketDisconnect:
            pass
        finally:
            if self.on_disconnect:
                ... # Run on disconnect logic here
```

**Comment by rafalp** [comment]
I'll assign this to myself so I can give this another look after #834/#802 gets done.

**Comment by rafalp** [comment]
The change I've mentioned two comments above is now part of #834. It'll be released with 0.15
